src/audio/mfcc/mfcc_init.c

```diff
@@ -0,0 +1,452 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+
+#include <sof/audio/mfcc/mfcc_comp.h>
+#include <sof/audio/component.h>
+#include <sof/audio/audio_stream.h>
+#include <sof/math/auditory.h>
+#include <sof/math/matrix.h>
+#include <sof/math/sqrt.h>
+#include <sof/math/trig.h>
+#include <sof/math/window.h>
+#include <sof/trace/trace.h>
+#include <user/mfcc.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define DEBUG_WITH_TESTBENCH
+
+#if defined(CONFIG_LIBRARY) && defined(DEBUG_WITH_TESTBENCH)
+#define DEBUGFILES
+#endif
+
+#ifdef DEBUGFILES
+#include <stdio.h>
+FILE *fh_window;
+FILE *fh_triangles;
+FILE *fh_dct;
+FILE *fh_lifter;
+#endif
+
+#define PI_Q29 Q_CONVERT_FLOAT(3.1415926536, 29)
+#define PI_Q23 Q_CONVERT_FLOAT(3.1415926536, 23)
+#define TWO_PI_Q24 Q_CONVERT_FLOAT(6.2831853072, 24)
+#define TWO_PI_Q23 Q_CONVERT_FLOAT(6.2831853072, 23)
+#define ONE_OVER_SQRT_TWO Q_CONVERT_FLOAT(0.7071067812, 31)
+#define HALF_Q1 Q_CONVERT_FLOAT(0.5, 1)
+#define ONE_Q9 Q_CONVERT_FLOAT(1, 9)
+#define TWO_Q29 Q_CONVERT_FLOAT(2, 29)
+
+#define DCT_MATRIX_MAX 42 /* Max matrix size where cos() argument fits Q7.24 */
+
+#ifdef MFCC_DEBUGFILES
+static void mfcc_init_debug_open(void)
+{
+	fh_window = fopen("fft_win.txt", "w");
+	fh_triangles = fopen("mel_triangles.txt", "w");
+	fh_dct = fopen("dct_matrix.txt", "w");
+	fh_lifter = fopen("lifter.txt", "w");
+}
+
+static void mfcc_init_debug_close(void)
+{
+	fclose(fh_window);
+	fclose(fh_triangles);
+	fclose(fh_dct);
+	fclose(fh_lifter);
+}
+#endif
+
+static void mfcc_init_buffer(struct mfcc_buffer *buf, int16_t *base, int size)
+{
+	buf->addr = base;
+	buf->end_addr = base + size;
+	buf->r_ptr = base;
+	buf->w_ptr = base;
+	buf->s_free = size;
+	buf->s_avail = 0;
+	buf->s_length = size;
+}
+
+static int mfcc_get_window(struct mfcc_state *state, enum fft_window_type name)
+{
+	struct mfcc_fft *fft = &state->fft;
+
+	switch (name) {
+	case RECTANGULAR_WINDOW:
+		win_rectangular_16b(state->window, fft->fft_size);
+		return 0;
+	case BLACKMAN_WINDOW:
+		win_blackman_16b(state->window, fft->fft_size, MFCC_BLACKMAN_A0);
+		return 0;
+	case HAMMING_WINDOW:
+		win_hamming_16b(state->window, fft->fft_size);
+		return 0;
+	case POVEY_WINDOW:
+		win_povey_16b(state->window, fft->fft_size);
+		return 0;
+
+	default:
+		return -EINVAL;
+	}
+}
+
+/* Return a matrix for discrete cosine transform (DCT). The DCT type is hard
+ * coded to DCT-II with orthogonal matrix. See
+ * https://en.wikipedia.org/wiki/Discrete_cosine_transform#DCT-II
+ *
+ * The size is (num_mel, num_ceps). Currently there are not other input parameters.
+ * TODO: Add option for no ortho norm.
+ */
+static int mfcc_get_dct_matrix(struct mfcc_dct *dct)
```

Comment:
This could be split into a separate library with unit tests since the DCT is generally usable and quite complex code here.

Reply:
done